### ISSUE TITLE
Issues combining `TypeGuard` and a `Concatenate`d `ParamSpec`

### ISSUE THREAD

**Issue by gjoseph92** [microsoft/pyright#2570]
**Describe the bug**
When using a `TypeGuard` on a function that's annotated with a `ParamSpec`, the function's concatenated parameters are lost within the narrowed branch.

**To Reproduce**
```python
from typing import Callable, TypeVar
from typing_extensions import Concatenate, ParamSpec, TypeGuard

Pi = ParamSpec("Pi")


def is_inty(f: Callable[Pi, object]) -> TypeGuard[Callable[Pi, int]]:
    return f.__name__.startswith("int")


Po = ParamSpec("Po")
T = TypeVar("T")


def outer(f: Callable[Concatenate[str, Po], T]) -> Callable[Po, T]:
    def inner(*args: Po.args, **kwargs: Po.kwargs) -> T:
        if is_inty(f):
            inty_f = f
            reveal_type(inty_f)
            # ^ `Callable[Po@outer, int]` - concat is missing?
            x = inty_f("foo", *args, **kwargs)
            #          ~~~~~
            # Expected 0 positional arguments
            return x + 1
        else:
            reveal_type(f)
            # ^ Callable[Concatenate[str, Po@outer], T@outer]
            return f("foo", *args, **kwargs)
            # ^ note that this type-checks fine

    return inner
```

**Expected behavior**
I'd _expect_ this to typecheck correctly. Maybe this is expected behavior though?

Within the `is_inty` block, you can see that `f`'s type has changed, and we've lost the `Concatenate`. It feels to me like concatenated params aren't getting passed through the `is_inty` TypeGuard correctly. That is, `is_inty` says it returns `Callable[Pi, int]`, but what's passed in for `Pi` is `Concatenate[int, Po]`. It's not tracking the fact that `Pi == Concatenate[int, Po]`?

**VS Code extension or command-line**
VS Code extension @ v1.1.187

**Additional context**
I originally ran into this trying to implement the suggestion in https://github.com/microsoft/pyright/discussions/1821#discussioncomment-690704. This example is a simplified version, where instead of `is_inty` narrowing to `int`, `is_coroutine_function` narrows to `Awaitable[T]`, but you end up with a similar problem.
xref https://github.com/microsoft/pyright/issues/2142

**Comment by erictraut** [comment]
Thanks for the bug report. As you pointed out, the `Concatenate` should be retained as part of the type guard. I've fixed that issue, and it will be addressed in the next release.

Even with that fix in place, the above code sample won't type check as is. The expression `x + 1` will evaluate to the type `int`, which isn't compatible with a return type of `T@outer`. The association between the original `T@outer` type variable and the type `int` is lost when passing the callable type through the type guard function. You can use a `cast` operator as a workaround.

```python
return cast(T, x + 1)
```

**Comment by gjoseph92** [comment]
Thanks for the quick fix @erictraut!

> The expression `x + 1` will evaluate to the type `int`, which isn't compatible with a return type of `T@outer`

I noticed this too, just forgot to remove it from my example. I would be happy to hear this is a bug related to `TypeGuard`; from https://github.com/python/mypy/issues/10302#issuecomment-816382424 I thought it might be expected behavior.

To me, this code without a custom `TypeGuard` (but a built-in type guard) fails in the same way:
```python
from typing import Callable, TypeVar

T = TypeVar("T")


def foo(f: Callable[[], T]) -> T:
    x = f()
    if isinstance(x, int):
        return x + 1
        #      ~~~~~
        # Expression of type "int" cannot be assigned to return type "T@foo"
    return x
```
Though maybe this is different from my example in a way I'm not understanding.

**Comment by erictraut** [comment]
Pyright is correct to flag this as an error. The expression `x + 1` is always type `int`. It's not type `T` (which could be a subclass of `int`). So it is a type violation to assign `x + 1` to type `T`. For comparison, mypy generates the same error here.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.188, which I just published. It will also be included in the next release of pylance.
